src/backend/executor/nodeHash.c

```diff
@@ -731,6 +731,10 @@ ExecHashIncreaseNumBatches(HashJoinTable hashtable)
 	if (oldnbatch > Min(INT_MAX / 2, MaxAllocSize / (sizeof(void *) * 2)))
 		return;
 
+	/* avoid repalloc_huge overflow on 32 bit systems */
+	if (stats && oldnbatch > MaxAllocHugeSize / (sizeof(HashJoinBatchStats) * 2))
+		return;
```

Comment:
1. Done.
2. Since `MaxAllocSize / ((sizeof(void *) * 2)` = `67 108 863`, `67 108 863 * 80` = `5 368 709 040`, which is slightly less than 5GB, so you are right.